### ISSUE TITLE
Disabling the capability of creating secrets

### ISSUE THREAD

**Issue by mshedsilegx** [Luzifer/ots#92]
Can you think of a way to prevent some users from creating secrets ?
I had some feedback on the tool and this question emerged. ie, having secret consumers only.

That seems difficult to me as we do not have roles in the application. Maybe a specific value encoded in the URL to just mask the new secret button ? Obviously, that's a poor implementation is visual only.

I would imagine a lot of interest of having the roles producer only, consumer only, producer/consumer.

**Comment by Luzifer** [comment]
This was once discussed in #20 and I started an implementation to disable the creation frontend (#21) but ditched it later on as the interest didn't seem too big for this.

If the interest in this is big enough, we could kind of revive that with the new customization options to just disable the creator frontend.

Of course this would mean you'd need to create secrets through the API and protect the API with something like [nginx-sso](https://github.com/Luzifer/nginx-sso) or similar ACL tooling capable of protecting some routes from access for users not in the right group / not logged into company SSO. Alternatively you could run two instances against the same Redis and use one to create - which would give you URLs not belonging to the consuming site - and one to consume…

On the other hand: What's the benefit of this? OTS is designed as a simple tool to let users share secrets without handing them over to the server in plain text as many other one-time-secret tools are doing and I'm a bit protective against feature-creep as I want to keep it as simple as possible.

If the thought is someone might overwhelm the storage: You can limit the body size of the route to 1K and call it a day…
If it's about someone storing "nasty" things: They are fully encrypted which **should** be not an issue legal-wise (I'm not a lawyer, this is not legal council!)…

Of course with now possible white-label instances it opens an attack vector for phishing as it seems the secret comes from the company but in fact was sent by a malicious third party… For this it might be sufficient to just go the "lets protect `/api/create` with ACL requiring company SSO" way: Anyone else trying to create a secret would just see "Well, something went wrong, oops." and wouldn't be able to create a secret…

Maybe it makes sense to indeed be able to disable the creation frontend for this but that makes it harder for employees allowed to create secrets to do so…

You see: There is a lot of thought about this and currently I'm not sure how and whether to proceed. Can you tell me more about the reasoning behind wanting to disable the creation? Maybe also whether one of the options above would already be a solution?

**Comment by mshedsilegx** [comment]
Yes, I agree, protecting /api/create might be sufficient. We are already frontending the app with nginx and auth.
So maybe just apply auth to /api/create, and let the rest open. What do you think ?

**Comment by HappyMan4** [comment]
For us it is sufficient to only allow internal networks to create secrets (via nginx allow rule). This is just to disallow the internet from creating secrets. If authentication is needed, as Luzifer already described, maybe sth like nginx-sso could be the way to go?

**Comment by Luzifer** [comment]
I think security-wise it's completely sufficient to limit the `/api/create` route. That way only authorized personell can create secrets in the instance.

UX wise it's kind of bad as the user is presented with the creation frontend but in the end cannot create a secret.

Another idea I had: We could add another API route which just returns an HTTP 200 which enables the creation frontend. If that `/api/writable` is blocked the frontend disables the creation. That way if secured by the same auth the creation frontend would be shown to those being logged in but disabled for those not being logged in (or having the correct permission)… 🤔

**Comment by mshedsilegx** [comment]
That's perfect IMO. And we would just have to restrict via auth/sso ```/api/writable``` in nginx, correct ?

**Comment by Luzifer** [comment]
Called it `/api/isWritable` now but yipp, correct… You'd restrict that endpoint exactly the same you are restricting `/api/create` so everyone being able to post to `create` gets an `HTTP 204` on `isWritable`

**Comment by mshedsilegx** [comment]
do we need to protect both /api/create and /api/isWritable, or just /api/isWritable ?

**Comment by Luzifer** [comment]
You could just protect `isWritable` but that would leave API access wide open. From SEC perspective you definitely want to close both endpoints.

**Comment by mshedsilegx** [comment]
Something is not right.
When the pasting the secret URL, the user is getting prompted as well for basic auth. Hitting cancel, and show me the secret works

**Comment by Luzifer** [comment]
Are you delivering the `WWW-Authenticate` header in the `isWritable` route? That might cause the `fetch` API to ask for a password. The `isWritable` should only respond non-204 without the Realm headers to disable the frontend.

**Comment by mshedsilegx** [comment]
The recipient can be internal or external, and the behavior should be consistent for both, ie no authentication required to consume the secret.

**Comment by Luzifer** [comment]
I haven't tested with basic auth, just with a redirect to a SSO login. It should also work with basic auth, just then you must not deliver the `WWW-Authenticate` header on the `isWritable` route.

My flow:
- User is not logged in
- `/api/isWritable` returns 403, no creator frontend
- Fetching the secret succeeds
- User logs in through login.example.com
- `/api/isWritable` returns 204, creator frontend is shown
- `/api/create` also is accepted as auth-cookie from login is available
- Fetching the secret succeeds

Using basic auth IMHO isn't really suited for SSO.

If you want to use basic auth you'd probably want to configure `/api/isWritable` just to return `403` without returning the `WWW-Authenticate` header. Probably the same on `/api/create`… And then add some `/login` route delivering the `WWW-Authenticate` header and a redirect to `/` on successful auth… Critical is that `/api/isWritable` **must not** cause the browser to show a login modal…

**Comment by mshedsilegx** [comment]
Appreciate the details. I'll fall back for now on protecting /api/create only, it works well for me with some customization of the message presented if the user escape the basic auth prompt. I'll use this feature when/if we implement a true SSO (okta) for the application. it will be very difficult, as we need external partners to also send us secrets, hence the lower possible denominator with basic auth.

**Comment by mshedsilegx** [comment]
If you agree, I think this issue can be closed.
